### PR TITLE
[GLUTEN-4454][CH] Support write mergetree data for CH backend

### LINKED ISSUE #4454: [CH] Support write mergetree data for CH backend
### Description

Support write mergetree data for CH backend:
1. create table for the mergetree data, and also support to config the `orderByKey`, `primaryKey` for CH;
2. now support to write to partition table and bucket table;
3. `insert overwrite` doesn't support now

Example:
```
CREATE TABLE IF NOT EXISTS lineitem_mergetree_partition
(
 l_orderkey      bigint,
 l_partkey       bigint,
 l_suppkey       bigint,
 l_linenumber    bigint,
 l_quantity      double,
 l_extendedprice double,
 l_discount      double,
 l_tax           double,
 l_returnflag    string,
 l_linestatus    string,
 l_commitdate    date,
 l_receiptdate   date,
 l_shipinstruct  string,
 l_shipmode      string,
 l_comment       string,
 l_shipdate      date
)
USING clickhouse
PARTITIONED BY (l_shipdate, l_returnflag)
TBLPROPERTIES (orderByKey='l_orderkey',
               primaryKey='l_orderkey')
LOCATION '$basePath/lineitem_mergetree_partition'
```

```
insert into table lineitem_mergetree_partition
   select
   l_orderkey      ,
 l_partkey       ,
 l_suppkey       ,
 l_linenumber    ,
 l_quantity      ,
 l_extendedprice ,
 l_discount      ,
 l_tax           ,
 l_returnflag    ,
 l_linestatus    ,
 l_commitdate    ,
 l_receiptdate   ,
 l_shipinstruct  ,
 l_shipmode      ,
 l_comment,
 l_shipdate
    from lineitem
   where l_shipdate BETWEEN date'1992-01-01' AND date'1992-06-10'
```

### PR DESCRIPTION
**Pull request by zzcclp**
## What changes were proposed in this pull request?

Support write mergetree data for CH backend:

1. create table for the mergetree data, and also support to config the orderByKey, primaryKey for CH;
2. now support to write to partition table and bucket table;
3. insert overwrite doesn't support now

Example:

```
CREATE TABLE IF NOT EXISTS lineitem_mergetree_partition
(
 l_orderkey      bigint,
 l_partkey       bigint,
 l_suppkey       bigint,
 l_linenumber    bigint,
 l_quantity      double,
 l_extendedprice double,
 l_discount      double,
 l_tax           double,
 l_returnflag    string,
 l_linestatus    string,
 l_commitdate    date,
 l_receiptdate   date,
 l_shipinstruct  string,
 l_shipmode      string,
 l_comment       string,
 l_shipdate      date
)
USING clickhouse
PARTITIONED BY (l_shipdate, l_returnflag)
TBLPROPERTIES (orderByKey='l_orderkey',
               primaryKey='l_orderkey')
LOCATION '$basePath/lineitem_mergetree_partition'
```

```
insert into table lineitem_mergetree_partition
   select
   l_orderkey      ,
 l_partkey       ,
 l_suppkey       ,
 l_linenumber    ,
 l_quantity      ,
 l_extendedprice ,
 l_discount      ,
 l_tax           ,
 l_returnflag    ,
 l_linestatus    ,
 l_commitdate    ,
 l_receiptdate   ,
 l_shipinstruct  ,
 l_shipmode      ,
 l_comment,
 l_shipdate
    from lineitem
   where l_shipdate BETWEEN date'1992-01-01' AND date'1992-06-10'
```

(Fixes: #4454)

Close #4454 

Co-authored-by: liuneng <1398775315@qq.com>
Co-authored-by: Shuai li <loneylee@live.cn>

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/benchmarks/ParquetReadBenchmark.scala [hunk 1]
```diff
@@ -82,7 +82,6 @@ object ParquetReadBenchmark extends SqlBasedBenchmark {
         .set("spark.io.compression.codec", "LZ4")
         .set("spark.gluten.sql.enable.native.validation", "false")
         .set("spark.gluten.sql.columnar.backend.ch.worker.id", "1")
-        .set("spark.gluten.sql.columnar.backend.ch.use.v2", "false")
```
Comment (baibaichen) [inline gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/benchmarks/ParquetReadBenchmark.scala:85]: 这个去掉的原因？

Comment (zzcclp) [reply gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/benchmarks/ParquetReadBenchmark.scala:85]: do not support DataSource V2 for the mergetree format

File: cpp-ch/local-engine/Builder/SerializedPlanBuilder.cpp [hunk 2]
```diff
@@ -190,13 +190,13 @@ SerializedPlanBuilder & SerializedPlanBuilder::readMergeTree(
...
 {
     substrait::Rel * rel = new substrait::Rel();
-    auto * read = rel->mutable_read();
-    read->mutable_extension_table()->mutable_detail()->set_value(local_engine::MergeTreeTable{
-        .database = database, .table = table, .relative_path = relative_path, .min_block = min_block, .max_block = max_block}
-                                                                     .toString());
-    read->set_allocated_base_schema(schema);
-    setInputToPrev(rel);
-    this->prev_rel = rel;
+    // auto * read = rel->mutable_read();
```
Comment (baibaichen) [inline cpp-ch/local-engine/Builder/SerializedPlanBuilder.cpp:193]: 这里似乎什么都没做？是不是可以删除了？

File: cpp-ch/local-engine/Storages/StorageMergeTreeFactory.h [hunk 3]
```diff
@@ -16,6 +16,7 @@
  */
 #pragma once
 #include <Storages/CustomStorageMergeTree.h>
+#include "Interpreters/MergeTreeTransaction.h"
```
Comment (baibaichen) [inline cpp-ch/local-engine/Storages/StorageMergeTreeFactory.h:19]: `include ""` => `include <>`

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/utils/MergeTreePartsPartitionsUtil.scala [hunk 4]
```diff
@@ -27,13 +29,60 @@ import org.apache.spark.sql.connector.read.InputPartition
...
+
+import scala.collection.JavaConverters._
 import scala.collection.mutable.ArrayBuffer
 
+// scalastyle:off argcount
 object MergeTreePartsPartitionsUtil extends Logging {
 
+  val DEFAULT_ORDER_BY_KEY = "tuple()"
+
+  def genOrderByAndPrimaryKeyStr(
```
Comment (binmahone) [inline backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/utils/MergeTreePartsPartitionsUtil.scala:46]: irrelevant to "MergeTreePartsPartitionsUtil"

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/utils/MergeTreePartsPartitionsUtil.scala [hunk 5]
```diff
@@ -27,13 +29,60 @@ import org.apache.spark.sql.connector.read.InputPartition
...
+        !orderByKey.equals(DEFAULT_ORDER_BY_KEY) && primaryKeyOption.isDefined &&
+        primaryKeyOption.get.nonEmpty
+      ) {
+        primaryKeyOption.get.mkString(",")
+      } else ""
+
+    (orderByKey, primaryKey)
+  }
+
+  def genNamedStructJson(tableSchema: StructType): String = {
```
Comment (binmahone) [inline backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/utils/MergeTreePartsPartitionsUtil.scala:65]: irrelevant to "MergeTreePartsPartitionsUtil"

File: cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    auto converter = ActionsDAG::makeConvertingActions(new_block.getColumnsWithTypeAndName(), header.getColumnsWithTypeAndName(), DB::ActionsDAG::MatchColumnsMode::Position);;
+    if (converter)
+    {
+        ExpressionActions do_convert = ExpressionActions(converter);
+        do_convert.execute(new_block);
+    }
+    auto res = squashing_transform->add(new_block);
+    if (res)
+    {
+        auto blocks_with_partition = MergeTreeDataWriter::splitBlockIntoParts(res, 10, metadata_snapshot, context);
```
Comment (binmahone) [inline cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp:54]: for us, the second param(max parts) should be 1?

File: cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        part_dir = fmt::format("{}_{:03d}", uuid, part_num);
+    }
+
+    String part_name = part_dir;
+
+    temp_part.temporary_directory_lock = storage.getTemporaryPartDirectoryHolder(part_dir);
+
+    auto indices = MergeTreeIndexFactory::instance().getMany(metadata_snapshot->getSecondaryIndices());
+
+    /// If we need to calculate some columns to sort.
```
Comment (binmahone) [inline cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp:123]: do we need to sort again? spark should have already sorted the input stream?

**Review by binmahone — COMMENTED** [review]
@liuneng1994

File: cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp (hunk 6 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp:54]: 不影响结果，不会发生实际的split

File: cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp (hunk 7 above)
Comment (liuneng1994) [reply cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp:123]: 这个二级索引的代码，目前还没用到，但是后面有二级索引的时候需要再进行排序

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/extension/ClickHouseAnalysis.scala [hunk 8]
```diff
@@ -39,7 +38,7 @@ class ClickHouseAnalysis(session: SparkSession, conf: SQLConf)
   override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsDown {
     // This rule falls back to V1 nodes according to 'spark.gluten.sql.columnar.backend.ch.use.v2'
     case dsv2 @ DataSourceV2Relation(tableV2: ClickHouseTableV2, _, _, _, options) =>
-      if (
+      /* if (
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/org/apache/spark/sql/extension/ClickHouseAnalysis.scala:41]: done

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/v2/clickhouse/metadata/AddFileTags.scala [hunk 9]
```diff
@@ -101,7 +110,8 @@ object AddFileTags {
       partition: String = "",
       defaultCompressionCodec: String = "LZ4",
       stats: String = "",
-      partitionValues: Map[String, String] = Map.empty[String, String]): AddFile = {
+      partitionValues: Map[String, String] = Map.empty[String, String],
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/v2/clickhouse/metadata/AddFileTags.scala:113]: add some comments for the current used field.

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/v1/clickhouse/MergeTreeFileFormatWriter.scala [hunk 10]
```diff
@@ -0,0 +1,681 @@
...
+  def setOptimisticTransactionHasWritten(txn: OptimisticTransaction): Unit = {
+    val txnRuntimeMirror = runtimeMirror(classOf[OptimisticTransaction].getClassLoader)
+    val txnInstanceMirror = txnRuntimeMirror.reflect(txn)
+    val txnHasWritten = typeOf[OptimisticTransaction].member(TermName("hasWritten_$eq")).asMethod
+    val txnHasWrittenMirror = txnInstanceMirror.reflectMethod(txnHasWritten)
+    txnHasWrittenMirror(true)
+  }
+
+  // scalastyle:off argcount
+  def writeFiles(
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/v1/clickhouse/MergeTreeFileFormatWriter.scala:233]: create a new `MergeTreeDeltaTxnWriter` object and move `writeFiles` to this object.

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/utils/MergeTreePartsPartitionsUtil.scala (hunk 5 above)
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/utils/MergeTreePartsPartitionsUtil.scala:65]: move to `ConverterUtils.convertNamedStructJson`

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/utils/MergeTreePartsPartitionsUtil.scala (hunk 4 above)
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/utils/MergeTreePartsPartitionsUtil.scala:46]: move to `MergeTreeDeltaUtil`

File: backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/v1/CHMergeTreeWriterInjects.scala [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+  // scalastyle:off argcount
+  override def createOutputWriter(
+      path: String,
+      database: String,
+      tableName: String,
+      orderByKeyOption: Option[Seq[String]],
+      primaryKeyOption: Option[Seq[String]],
+      partitionColumns: Seq[String],
+      bucketSpec: Option[BucketSpec],
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/org/apache/spark/sql/execution/datasources/v1/CHMergeTreeWriterInjects.scala:69]: removed

File: backends-clickhouse/src/main/scala/io/glutenproject/execution/GlutenMergeTreePartition.scala [hunk 12]
```diff
@@ -18,14 +18,29 @@ package io.glutenproject.execution
 
 import io.glutenproject.substrait.plan.PlanBuilder
 
+case class MergeTreePartitionedFile(
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply backends-clickhouse/src/main/scala/io/glutenproject/execution/GlutenMergeTreePartition.scala:21]: rename to `MergeTreePartSplit`

File: cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+{
+
+Block removeColumnSuffix(const DB::Block & block)
+{
+    ColumnsWithTypeAndName columns;
+    for (int i = 0; i < block.columns(); ++i)
+    {
+        auto name = block.getByPosition(i).name;
+        Poco::StringTokenizer splits(name, "#");
+        ;
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply cpp-ch/local-engine/Storages/Mergetree/SparkMergeTreeWriter.cpp:34]: done

File: cpp-ch/local-engine/Common/MergeTreeTool.cpp [hunk 14]
```diff
@@ -15,16 +15,19 @@
...
 #include <IO/WriteHelpers.h>
+#include <Storages/MergeTree/IMergeTreeDataPart.h>
+#include <google/protobuf/util/json_util.h>
 
 using namespace DB;
 
 namespace local_engine
 {
-std::shared_ptr<DB::StorageInMemoryMetadata> buildMetaData(DB::NamesAndTypesList columns, ContextPtr context)
+std::shared_ptr<DB::StorageInMemoryMetadata> buildMetaData(DB::NamesAndTypesList columns, ContextPtr context, MergeTreeTable table)
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply cpp-ch/local-engine/Common/MergeTreeTool.cpp:30]: done

File: cpp-ch/local-engine/local_engine_jni.cpp [hunk 15]
```diff
@@ -984,8 +1042,20 @@ JNIEXPORT void Java_org_apache_spark_sql_execution_datasources_CHDatasourceJniWr
     LOCAL_ENGINE_JNI_METHOD_END(env, )
 }
 
+JNIEXPORT jstring Java_org_apache_spark_sql_execution_datasources_CHDatasourceJniWrapper_closeMergeTreeWriter(JNIEnv * env, jobject, jlong instanceId)
```
[the remark this answers is not in the document]
Comment (zzcclp) [reply cpp-ch/local-engine/local_engine_jni.cpp:1045]: done

**Review by binmahone — DISMISSED** [review]

**Review by binmahone — APPROVED** [review]

**Pull request merged**